### PR TITLE
query: fix querying with interleaved data

### PR DESCRIPTION
**Pull request by GiedriusS**
Fix querying when some data has been pushed down and some - hasn't.
Since `max_over_time` and `min_over_time` remove `__name__` from the
results either way, let's do that inside Select() to have a unified form
of data.

Add test to cover this case.

Without this code, the test fails:

```
level=error ts=2022-01-06T16:28:37.959956685Z msg="function failed. Retrying in next tick" err="read query instant response: expected 2xx response, got 422. Body: {\"status\":\"error\",\"errorType\":\"execution\",\"error\":\"vector cannot contain metrics with the same labelset\"}\n"
```

Closes https://github.com/thanos-io/thanos/issues/5033.

### CODE REVIEW

**Comment by GiedriusS** [comment]
cc @fpetkovski

**Review by fpetkovski — DISMISSED** [review]

File: test/e2e/query_test.go [hunk 1]
```diff
@@ -631,6 +638,93 @@ func newSample(s fakeMetricSample) model.Sample {
...
+// min_over_time.
+func TestSidecarStorePushdown(t *testing.T) {
+	t.Parallel()
+
+	// Build up.
+	e, err := e2e.NewDockerEnvironment("e2e_sidecar_store_pushdown")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	prom1, sidecar1, err := e2ethanos.NewPrometheusWithSidecar(e, "p1", defaultPromConfig("p1", 0, "", ""), "", e2ethanos.DefaultPrometheusImage(), "remote-write-receiver")
```
Comment (yeya24) [inline test/e2e/query_test.go:652]: Is it intended to enable remote write receiver?

File: test/e2e/query_test.go [hunk 2]
```diff
@@ -631,6 +638,93 @@ func newSample(s fakeMetricSample) model.Sample {
...
+	queryAndAssertSeries(t, ctx, q.Endpoint("http"), func() string {
+		return "max_over_time(my_fake_metric[2h])"
+	}, time.Now, promclient.QueryOptions{
+		Deduplicate: true,
+	}, []model.Metric{
+		{
+			"instance":   "foo",
+			"prometheus": "p1",
+		},
+	})
```
Comment (yeya24) [inline test/e2e/query_test.go:725]: Let's add another query for `min_over_time` as well?

File: test/e2e/query_test.go (hunk 1 above)
Comment (fpetkovski) [reply test/e2e/query_test.go:652]: The test synthesizes samples by remote writing them, which is why the receiver feature is needed.

File: test/e2e/query_test.go (hunk 2 above)
Comment (GiedriusS) [reply test/e2e/query_test.go:726]: Added tests for `min_over_time` & some other tricky cases: https://github.com/thanos-io/thanos/pull/5035/commits/203f449843681d05157030bec3cfa094f52a9fe7.

File: test/e2e/query_test.go (hunk 1 above)
Comment (yeya24) [reply test/e2e/query_test.go:653]: Thanks @fpetkovski. I thought it was using data from the block only. Didn't notice the remote write part.

**Review by yeya24 — APPROVED** [review]
LGTM

File: pkg/query/querier.go [hunk 3]
```diff
@@ -310,6 +311,22 @@ func (q *querier) selectFn(ctx context.Context, hints *storage.SelectHints, ms .
 		warns = append(warns, errors.New(w))
 	}
 
+	// Delete the metric's name from the result because that's what the
+	// PromQL does either way and we want our iterator to work with data
+	// that was either pushed down or not.
+	if q.enableQueryPushdown && (hints.Func == "max_over_time" || hints.Func == "min_over_time") {
```
Comment (yeya24) [inline pkg/query/querier.go:317]: Just curious why promql does this. Isn't it a bug?

Comment (GiedriusS) [reply pkg/query/querier.go:317]: Nope, it is a feature. Think of these functions like they are calculating the maximum over _all_ input, not just for each unique labelset. I think this is where it happens: https://github.com/prometheus/prometheus/blob/d677aa4b29a8a0cf9f61af04bbf5bfdce893cf23/promql/engine.go#L1304-L1310.

Comment (GiedriusS) [reply pkg/query/querier.go:317]: Once more functions are being pushed down, we will need to improve the logic here. Probably we will need to refactor this into a separate struct? I don't know yet

**Pull request merged**